adapters/lora/src/main/java/org/eclipse/hono/adapter/lora/LoraCommandSubscriptions.java

```diff
@@ -0,0 +1,178 @@
+
+    /**
+     * Creates a new Lora command subscriptions class.
+     *
+     * @param vertx The Vert.x instance to use.
+     */
+    public LoraCommandSubscriptions(final Vertx vertx) {
+        this(vertx, null);
+    }
+
+    /**
+     * Creates a new Lora command subscriptions class.
+     *
+     * @param vertx The Vert.x instance to use.
+     * @param tracer The tracer instance.
+     */
+    public LoraCommandSubscriptions(final Vertx vertx, final Tracer tracer) {
+        this.commandSubscriptions = new ConcurrentHashMap<>();
+        this.vertx = vertx;
+        this.tracer = tracer == null ? NoopTracerFactory.create() : tracer;
+    }
+
+    /**
+     * Subscribes for tenant diasbled/deleted events, in order to close the consumer objects associated with them later.
+     */
+    public void init() {
+        NotificationEventBusSupport.registerConsumer(vertx, TenantChangeNotification.TYPE,
+                notification -> {
+                    if (LifecycleChange.DELETE.equals(notification.getChange())
+                            || (LifecycleChange.UPDATE.equals(notification.getChange())
+                                    && !notification.isTenantEnabled())) {
+                        closeConsumersForTenant(notification.getTenantId());
+                    }
+                });
+
+        NotificationEventBusSupport.registerConsumer(vertx, DeviceChangeNotification.TYPE,
+                notification -> {
+                    if (LifecycleChange.DELETE.equals(notification.getChange())) {
+                        closeConsumersForGateway(notification.getTenantId(), notification.getDeviceId());
+                    }
+                });
+    }
+
+    @SuppressWarnings("rawtypes")
+    private void closeConsumersForGateway(final String tenantId, final String gatewayId) {
+        if (!commandSubscriptions.entrySet().stream().anyMatch(e -> e.getKey().getDeviceId().equals(gatewayId))) {
+            return;
+        }
+
+        final Span span = TracingHelper
+                .buildSpan(tracer, null, "gateway deleted, corresponding command consumer will be closed\"",
+                        getClass().getSimpleName())
+                .withTag(Tags.SPAN_KIND.getKey(), Tags.SPAN_KIND_CLIENT)
+                .start();
+        TracingHelper.setDeviceTags(span, tenantId, gatewayId);
+        final SubscriptionKey key = new SubscriptionKey(tenantId, gatewayId);
+        final TriTuple<CommandConsumer, LoraProvider, Context> subscription = commandSubscriptions.get(key);
+        final List<Future> consumerCloseFutures = Collections.synchronizedList(new ArrayList<>());
+        final Context ctx = subscription.three();
+        ctx.runOnContext(x -> {
+            final CommandConsumer commandConsumer = subscription.one();
+            consumerCloseFutures.add(commandConsumer.close(null));
+        });
+        CompositeFuture.join(consumerCloseFutures)
+                .onComplete(x -> commandSubscriptions.remove(key))
```

Comment:
I think there is no need to wait for the `consumerCloseFutures` here - the `commandSubscriptions` can be removed right away (see other comment).